### ISSUE TITLE
VFS NFS 4+, simultaneous read/writes sometimes get discarded?

### ISSUE THREAD

**Issue by wongma7** [nfs-ganesha/nfs-ganesha#396]
If I have process 1 opening and writing to a file and process 2 opening and reading from it at roughly the same time sometimes the writes from process 1 don't "commit." What should happen? Not sure how caching/locking/etc. should work, all I know is that when mounting with nfsvers=3 or when using kernel nfs, the write always goes through...

Reproduce (crude method...), ganesha 2.7.1:
- Mount something at `/tmp/nfs`

write.c
```c
#include <stdio.h>

main() {
   FILE *fp;

   fp = fopen("/tmp/nfs/test.txt", "w+");
   fputs("testing\n", fp);
   fclose(fp);
}
```
read.c
```c
#include <stdio.h>

main() {

   FILE *fp;
   char buff[255];

   fp = fopen("/tmp/nfs/test.txt", "r");
   fgets(buff, 255, (FILE*)fp);
   printf("%s", buff);
   fclose(fp);
}
```

- Run:
$ rm /tmp/nfs/test.txt; ./write& ./read; sleep 1; ./read

Output with ganesha mount:
Most of the time file is empty:
```
[1] 15748
[1]+  Done                    ./write
```
Sometimes file gets stuff written to it:
```
[1] 15753
[1]+  Done                    ./write
Segmentation fault (core dumped)
testing
```

Output with kernel mount or nfsvers=3 mount:
File always gets written to:
```
[1] 15938
testing
[1]+  Done                    ./write
testing
```

**Comment by dang** [comment]
NFSv4 on Linux has what's called "close to open" semantics.  This means that changes made while the file are open don't necessarily show up until the file is closed.  The next open after the close will have all the committed changes.  See here for more info:
http://nfs.sourceforge.net/#faq_a8

In this case, since you're running the programs in the background, the timing will be random, based on the scheduler and the NIC and the switch and the Ganesha server and what else is running on the client, and...

The only way to ensure that something you've written shows up to another process is for both of them to use locking, or for them to use some form of IPC to communicate when one has closed the file so the other can open it.

In short, I don't think this is a bug in Ganesha, but is NFSv4 working as it should.

Not that NFSv3 doesn't have OPEN or CLOSE ops on the network.  Instead, each read or write does an open+read+close or an open+write+close on the server.  This make the window much smaller between when the write is done and the read can see it, but it's less efficient.

**Comment by wongma7** [comment]
Thanks, I understand better now. To be clear, I am seeing that the writer's changes often *never* get committed. It makes sense if the reader with the file open concurrently sees the changes only randomly if it happens to open the file after the writer has closed the file, but I hoped subsequent readers would always be able to see the changes, and instead the changes are often discarded forever.

I guess then my bug report is more like an enhancement request to make the behaviour more like kernel nfs where it seems (in my brief testing) that the writer's changes never disappear, even if the nfs spec doesn't say anything about guaranteeing this

**Comment by dang** [comment]
Hmm...  If it never shows up, that's a bug.  I'll take a look.

**Comment by ffilz** [comment]
Closing this due to age, feel free to re-open if it's still of interest and you have explored using a currently supported version of Ganesha (V2.8 or V3 at this time).
